### ISSUE TITLE
Packages in `jspm.config.js` are not respected

### ISSUE THREAD

**Issue by ghost** [skorlir/ava-jspm-loader#3]
​

**Comment by fluffywaffles** [comment]
For later reference:

The problem is specifically with package dependencies specified in the "packages" section of system config but not in the "map" section or in your jspm dependencies in package.json.

**Comment by fluffywaffles** [comment]
I am able to work around this currently by just doing `jspm i fbjs`.

I don't consider it a long-term solution, but `npm test` does run without errors if you do that.

**Comment by fluffywaffles** [comment]
The trouble with this error is: If I try to replicate it in the node console, it fails. SystemJS doesn't load `fbjs` except indirectly if you try to load `react`.

Try:

``` bash
$ node
> const jspm = require('jspm')
undefined
> const System = jspm.Loader()
undefined
> System.import('react').then(console.log.bind(console))
/... logs the contents of the react module
> System.import('react').then(function (React) {
.. class Example extends React.Component {}
.. console.log(Example)
..})
[Function: Example]
undefined
> System.import('fbjs').then(console.log.bind(console)).catch(console.log.bind(console))
/... logs an error - fbjs cannot be found
```

If SystemJS cannot load it, then neither can ava-jspm-loader.

I'm not sure by what convoluted series of transformations the resulting babel-emitted code is deciding to try to `require('fbjs/lib/immutable')`, but SystemJS won't load that dependency (it claims it cannot find it) and so it forwards the request on to the Node Module loader, which of course cannot find it and throws an exception.

**Comment by fluffywaffles** [comment]
Not so much "won't fix" as it is "can't fix." If SystemJS won't load it, neither can ava-jspm-loader.

The best solution I have for you is the workaround I mentioned above: directly install fbjs so that the SystemJS loader knows how to resolve an import of it.

**Comment by kdex** [comment]
Huh. Installing `fbjs` directly does indeed make the error about it go away, but then I'm hit with dozens of other modules that can't be resolved which need to be installed manually via `jspm i`.

I observed that SystemJS will resolve React and the other dozens just fine when I change `ava-jspm-loader` to [that](https://github.com/avajs/ava/pull/941#issuecomment-229185405) and start my tests with something like:

```js
global.cache = (await System.import("client/cache")).default;
```

That's resolved just fine by SystemJS; no need to `jspm i` anything twice.

**Comment by fluffywaffles** [comment]
You're welcome to take that approach.

The bug regarding `fbjs` and other dependencies is almost certainly caused by the way the babel presets and plugins traverse your imports and transform their source. Like I said, SystemJS has no difficulty resolving imports of modules you have installed directly, even if they have indirectly installed dependencies. By using the SystemJS API directly, your babel transforms won't apply to those files because transforms won't interpret "System.import" as being an import. So it may cause other bugs down the line, for instance with coverage reporting. But I don't know. I haven't used that approach.

Basically, you are removing the loader from your workflow and using the SystemJS API directly instead. Pick your poison. I personally prefer to configure my projects in such a way that I don't have to write any boilerplate code in my tests, which is why I wrote this. A better solution would be to run tests via JSPM instead of using a separate, AVA-managed babel workflow for transpilation, but alas there's no obvious or easy way of doing that unless (until?) someone writes a test runner built on SystemJS.

**Comment by kdex** [comment]
I understand; so we'll probably has to close this as "can't fix" and just require that some `jspm` dependencies must be installed more than once. Both ways have obvious (dis)advantages. :confused:
